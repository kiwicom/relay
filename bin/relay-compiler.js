#!/usr/bin/env node

// @flow

// TODO: DRY monorepo-utils/bin/monorepo-babel-node-runner.js
// This is here to make this `bin` available directly from our monorepo without transpiling it.
require('@babel/register')({
  ignore: [/node_modules\/(?!@kiwicom)/],
  rootMode: 'upward',
});

const program = require('commander');
const { invariant } = require('@kiwicom/js');

const compiler = require('../src/compiler').default;

program
  .option('--src <src>')
  .option('--schema <schema>')
  .parse(process.argv);

invariant(program.src, 'Option --src is required.');
invariant(program.schema, 'Option --schema is required.');

compiler({
  src: program.src,
  schema: program.schema,
});
