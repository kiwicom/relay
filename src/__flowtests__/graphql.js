// @flow

import { graphql } from '../index';

module.exports = {
  correctUsage() {
    return graphql`correct usage`; // eslint-disable-line relay/graphql-syntax
  },
  correctUsageUnusual() {
    return graphql(['correct']);
  },
  incorrectUsage() {
    // $FlowExpectedError: input should be only string
    return graphql([1, 2]);
  },
};
