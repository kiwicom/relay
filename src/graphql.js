// @flow

import { graphql as _graphql } from 'react-relay';

import type { GraphQLTaggedNode } from './types.flow';

export default function graphql(
  strings: $ReadOnlyArray<string>,
): GraphQLTaggedNode {
  return _graphql(strings);
}
