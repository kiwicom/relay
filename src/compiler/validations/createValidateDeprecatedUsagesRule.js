// @flow

import { ConsoleReporter } from 'relay-compiler';
import { findDeprecatedUsages, type DocumentNode, type ValidationContext } from 'graphql';
import { sprintf } from '@kiwicom/js';

export default function createValidateDeprecatedUsagesRule(reporter: ConsoleReporter) {
  return function validateDeprecatedUsagesRule(validationContext: ValidationContext) {
    const schema = validationContext.getSchema();
    return {
      Document(node: DocumentNode): void {
        const deprecatedUsages = findDeprecatedUsages(schema, node);
        const errors = new Map(); // [error message, number of occurrences]
        deprecatedUsages.forEach(({ message }) => {
          if (errors.has(message)) {
            errors.set(message, errors.get(message) + 1);
          } else {
            errors.set(message, 1);
          }
        });
        errors.forEach((occurrences, errorMessage) => {
          reporter.reportMessage(sprintf('ğŸ’” %sx %s', occurrences, errorMessage));
        });
      },
    };
  };
}
