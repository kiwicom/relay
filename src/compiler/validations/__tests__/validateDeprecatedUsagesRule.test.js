// @flow

import os from 'os';
import path from 'path';
import { parse, validate } from 'graphql';
import { ConsoleReporter } from 'relay-compiler';
import { generateTestsFromFixtures } from '@kiwicom/test-utils';

import testSchema from './testSchema';
import createValidateDeprecatedUsagesRule from '../createValidateDeprecatedUsagesRule';

generateTestsFromFixtures(path.join(__dirname, 'fixtures', 'findDeprecatedUsagesRule'), query => {
  const documentAST = parse(query);
  const reporter = new ConsoleReporter({ quiet: true });
  const spy = jest.spyOn(reporter, 'reportMessage');
  const validateDeprecatedUsagesRule = createValidateDeprecatedUsagesRule(reporter);
  validate(testSchema, documentAST, [validateDeprecatedUsagesRule]);
  // $FlowIssue: https://github.com/facebook/flow/issues/7397
  const spyCalls = spy.mock.calls.flat(Infinity);
  return spyCalls.length === 0 ? 'PASSED âœ…' : spyCalls.join(os.EOL);
});
