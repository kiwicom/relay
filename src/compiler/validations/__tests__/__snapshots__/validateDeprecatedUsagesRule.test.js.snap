// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: deprecatedAlias.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  alias: fieldDeprecated1
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
ðŸ’” 1x The field QueryRoot.fieldDeprecated1 is deprecated. Use field "fieldString" instead.
`;

exports[`matches expected output: multipleDeprecated.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  fieldDeprecated1
  x: fieldDeprecated2
  y: fieldDeprecated2
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
ðŸ’” 1x The field QueryRoot.fieldDeprecated1 is deprecated. Use field "fieldString" instead.
ðŸ’” 2x The field QueryRoot.fieldDeprecated2 is deprecated. Use field "fieldString" instead.
`;

exports[`matches expected output: simple.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  fieldString
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
PASSED âœ…
`;

exports[`matches expected output: simpleWithDeprecated.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  fieldString
  fieldDeprecated1
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
ðŸ’” 1x The field QueryRoot.fieldDeprecated1 is deprecated. Use field "fieldString" instead.
`;
